### PR TITLE
[SW-228109][v0] [llama4 ]Llama 4 support for vLLM fork

### PR DESCRIPTION
**Pull request by leopck**
Initial support for Llama 4 on vLLM Fork cherry picked from https://github.com/HabanaAI/vllm-fork/commits/llama4

https://jira.habana-labs.com/browse/SW-228109

acc is now Good

vllm (pretrained=.../Llama-4-Maverick-17B-128E-Instruct/,tensor_parallel_size=8,max_model_len=4096,max_num_seqs=1024,gpu_memory_utilization=0.8,use_v2_block_manager=True,dtype=bfloat16,max_gen_toks=2048,disable_log_stats=False,enable_expert_parallel=True), gen_kwargs: (None), limit: 0.1, num_fewshot: 0, batch_size: auto
|       Tasks       |Version|    Filter    |n-shot|  Metric   |   |Value |   |Stderr|
|-------------------|------:|--------------|-----:|-----------|---|-----:|---|-----:|
|mmlu_pro           |      2|custom-extract|      |exact_match|↑  |0.8040|±  |0.0111|
| - biology         |      1|custom-extract|     0|exact_match|↑  |0.9167|±  |0.0328|
| - business        |      1|custom-extract|     0|exact_match|↑  |0.8481|±  |0.0406|
| - chemistry       |      1|custom-extract|     0|exact_match|↑  |0.8421|±  |0.0343|
| - computer_science|      1|custom-extract|     0|exact_match|↑  |0.8293|±  |0.0595|
| - economics       |      1|custom-extract|     0|exact_match|↑  |0.8471|±  |0.0393|
| - engineering     |      1|custom-extract|     0|exact_match|↑  |0.7629|±  |0.0434|
| - health          |      1|custom-extract|     0|exact_match|↑  |0.7195|±  |0.0499|
| - history         |      1|custom-extract|     0|exact_match|↑  |0.5897|±  |0.0798|
| - law             |      1|custom-extract|     0|exact_match|↑  |0.6216|±  |0.0462|
| - math            |      1|custom-extract|     0|exact_match|↑  |0.9412|±  |0.0203|
| - other           |      1|custom-extract|     0|exact_match|↑  |0.6989|±  |0.0478|
| - philosophy      |      1|custom-extract|     0|exact_match|↑  |0.7000|±  |0.0655|
| - physics         |      1|custom-extract|     0|exact_match|↑  |0.9077|±  |0.0255|
| - psychology      |      1|custom-extract|     0|exact_match|↑  |0.8250|±  |0.0427|

| Groups |Version|    Filter    |n-shot|  Metric   |   |Value|   |Stderr|
|--------|------:|--------------|------|-----------|---|----:|---|-----:|
|mmlu_pro|      2|custom-extract|      |exact_match|↑  |0.804|±  |0.0111|

vllm-vlm (pretrained=.../Llama-4-Maverick-17B-128E-Instruct/,enforce_eager=True,tensor_parallel_size=8,max_model_len=4096,max_num_seqs=64,gpu_memory_utilization=0.9,use_v2_block_manager=True,dtype=bfloat16,max_gen_toks=2048,disable_log_stats=False,enable_expert_parallel=True), gen_kwargs: (None), limit: None, num_fewshot: 0, batch_size: auto
| Tasks |Version|Filter|n-shot|     Metric      |   |Value |   |Stderr|
|-------|------:|------|-----:|-----------------|---|-----:|---|-----:|
|chartqa|      0|none  |     0|anywhere_accuracy|↑  |0.8884|±  |0.0063|
|       |       |none  |     0|exact_match      |↑  |0.6208|±  |0.0097|
|       |       |none  |     0|relaxed_accuracy |↑  |0.8868|±  |0.0063|

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -139,6 +139,12 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
                 )
             else:
                 raise NotImplementedError("CPU MOE only supports x86 arch.")
+        if current_platform.is_hpu():
+            for expert_id in range(layer.local_num_experts):
+                layer.moe_op.w13_list[expert_id].set_weight(
+                    layer.w13_weight.data[expert_id])
+                layer.moe_op.w2_list[expert_id].set_weight(
+                    layer.w2_weight.data[expert_id])
```
Comment (xuechendi) [inline vllm/model_executor/layers/fused_moe/layer.py:147]: Moved weight setting from _load_w13 and _load_w2 to process_weights_after_loading.
This works better for both llama4, qwen3 and deepseek

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -642,9 +648,6 @@ def _load_w13(self,
             expert_data = expert_data.narrow(shard_dim, shard_size, shard_size)
         expert_data.copy_(loaded_weight)
 
-        if is_hpu and isinstance(self.quant_method, UnquantizedFusedMoEMethod):
-            self.moe_op.w13_list[expert_id].set_weight(orig_exp_data)
```
Comment (xuechendi) [inline vllm/model_executor/layers/fused_moe/layer.py:646]: Since llama4 is loading full instead of loading expert 1 by 1, set weight for VllmMixureOfExpert weight in this function is a little tricky.
Because of that, we moved the function to process_weights_after_loading.

File: vllm/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -2727,6 +2725,38 @@ def execute_model(
...
+                                result.extend(fixed)
+                            else:
+                                result.append(fixed)
+                        return result
+                    else:
+                        return None
+
+                model_input.multi_modal_kwargs[
+                    'embed_is_patch'] = fix_embed_is_patch(
+                        model_input.multi_modal_kwargs['embed_is_patch'])
```
Comment (xuechendi) [inline vllm/worker/hpu_model_runner.py:2759]: necessary changes to unify multiple images format.

**Review by xuechendi — APPROVED** [review]
